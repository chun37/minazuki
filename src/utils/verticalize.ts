const verticalAndHorizontalCharactorMap: {
  targets: string[];
  vertical: string;
}[] = [
  {
    targets: ['ー'],
    vertical: '｜',
  },
  {
    targets: ['（', '('],
    vertical: '︵',
  },
  {
    targets: [')', '）'],
    vertical: '︶',
  },
  {
    targets: ['「', '｢'],
    vertical: '﹁',
  },
  {
    targets: ['｣', '」'],
    vertical: '﹂',
  },
  {
    targets: ['『'],
    vertical: '﹃',
  },
  {
    targets: ['』'],
    vertical: '﹄',
  },
  {
    targets: ['[', '［'],
    vertical: '﹇',
  },
  {
    targets: [']', '］'],
    vertical: '﹈',
  },
  {
    targets: ['{', '｛'],
    vertical: '︷',
  },
  {
    targets: ['}', '｝'],
    vertical: '︸',
  },
  {
    targets: ['〔', '❲'],
    vertical: '︹',
  },
  {
    targets: ['〕', '❳'],
    vertical: '︺',
  },
  {
    targets: ['〈', '〈'],
    vertical: '︿',
  },
  {
    targets: ['〉', '〉'],
    vertical: '﹀',
  },
  {
    targets: ['《', '⟪'],
    vertical: '︽',
  },
  {
    targets: ['》', '⟫'],
    vertical: '︾',
  },
  {
    targets: ['【'],
    vertical: '︻',
  },
  {
    targets: ['】'],
    vertical: '︼',
  },
  {
    targets: ['〖'],
    vertical: '︗',
  },
  {
    targets: ['〗'],
    vertical: '︘',
  },
];

/**
 * Transform `a-z` to `ａ-ｚ` style
 * @param text text to transform
 * @return transformed text
 */
export const transformLatinToFullWidth = (text: string) => {
  return text.replace(/[A-Za-z0-9]/g, tmpText =>
    String.fromCharCode(tmpText.charCodeAt(0) + 0xfee0),
  );
};

/**
 * Transform specific charactors to mapped vartical style charactor
 * @param text text to transfrom
 * @return transformed text
 */
export const transformHorizontalToVerticalUnicode = (text: string) => {
  let newText = text;

  verticalAndHorizontalCharactorMap.forEach(({ targets, vertical }) => {
    targets.forEach(target => {
      newText = newText.replace(target, vertical);
    });
  });

  return newText;
};

/**
 * Verticalize Japanese text from horizontal
 * @param text Target text to transform
 * @return Varticalized text
 */
export function verticalize(text: string): string {
  const lines = text
    .trim()
    .split('\n')
    .map(line => line.split(''))
    .reverse();
  const whiteSpace = '　';
  let verticalizedText = '';
  let index = 0;

  while (true) {
    const charAtLine = lines.map(line =>
      line[index] ? line[index] : whiteSpace,
    );

    if (charAtLine.join('') === whiteSpace.repeat(lines.length)) {
      break;
    }

    verticalizedText += `${charAtLine.join('')}\n`;
    index += 1;
  }

  // Avoid trimming
  verticalizedText = verticalizedText.replace(/^\s+?/, `.${whiteSpace}`);

  // Transforming to appropriate unicodes
  verticalizedText = transformHorizontalToVerticalUnicode(
    transformLatinToFullWidth(verticalizedText),
  );

  return verticalizedText;
}
