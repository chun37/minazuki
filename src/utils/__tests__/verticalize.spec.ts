import verticalize from '../verticalize';

describe('verticalize', () => {
  it ('verticalizes text', () => {
    const text = `
      Discord（ディスコード）とは、Windows、macOS、Android、iOS、Linux、Webブラウザで動作する、
      ゲーマー向けの音声通話・VoIPフリーソフトウェア。
      2018年5月16日時点で、ユーザー数が1億3千万人を上回っている。
    `;

    const result = verticalize(text);

    expect(result).toMatchSnapshot();
  });
});
