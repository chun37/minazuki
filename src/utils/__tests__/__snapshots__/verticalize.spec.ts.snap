// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verticalize verticalizes text 1`] = `
".　 Ｄ
  ｉ
  ｓ
  ｃ
  ｏ
  ｒ
２ゲｄ
０ー（
１マデ
８ーィ
年向ス
５けコ
月のー
１音ド
６声）
日通と
時話は
点・、
でＶＷ
、ｏｉ
ユＩｎ
ーＰｄ
ザフｏ
ーリｗ
数ーｓ
がソ、
１フｍ
億トａ
３ウｃ
千ェＯ
万アＳ
人。、
を　Ａ
上　ｎ
回　ｄ
っ　ｒ
て　ｏ
い　ｉ
る　ｄ
。　、
　　ｉ
　　Ｏ
　　Ｓ
　　、
　　Ｌ
　　ｉ
　　ｎ
　　ｕ
　　ｘ
　　、
　　Ｗ
　　ｅ
　　ｂ
　　ブ
　　ラ
　　ウ
　　ザ
　　で
　　動
　　作
　　す
　　る
　　、
"
`;
